### ISSUE TITLE
[HttpListener] [Porting] HttpListener.GetContext on Windows throws an NRE for NTLM and Negotiate authentication

### ISSUE THREAD

**Issue by hughbe** [dotnet/corefx#20096]
The following test gives an example of this (I'm currently writing tests for HttpListener authentication):

```cs
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace System.Net.Tests
{
    public class AuthenticationTests : IDisposable
    {
        private HttpListenerFactory _factory;
        private HttpListener _listener;

        public AuthenticationTests()
        {
            _factory = new HttpListenerFactory();
            _listener = _factory.GetListener();
        }

        public void Dispose() => _factory.Dispose();

        [Fact]
        public async Task NtlmAuthentication_NoRequestHeaders_ReturnsForbiddenStatusCode()
        {
            _listener.AuthenticationSchemes = AuthenticationSchemes.Ntlm;

            using (var client = new HttpClient())
            {
                HttpResponseMessage message = await AuthenticationFailure(client, HttpStatusCode.Unauthorized);
                Assert.Equal("NTLM", message.Headers.WwwAuthenticate.ToString());
            }
        }
        
        public async Task<HttpResponseMessage> AuthenticationFailure(HttpClient client, HttpStatusCode errorCode)
        {
            Task<HttpResponseMessage> clientTask = client.GetAsync(_factory.ListeningUrl);

            // The server task will hang forever if it is not cancelled.
            var tokenSource = new CancellationTokenSource();
            Task<HttpListenerContext> serverTask = Task.Run(() => _listener.GetContext(), tokenSource.Token);

            // The client task should complete first - the server should send a 401 response.
            Task resultTask = await Task.WhenAny(clientTask, serverTask);
            tokenSource.Cancel();
            if (resultTask == serverTask)
            {
                await serverTask;
            }

            Assert.Same(clientTask, resultTask);

            Assert.Equal(errorCode, clientTask.Result.StatusCode);
            return clientTask.Result;
        }
    }
}
```

This does not fail on desktop. The netcoreapp stack trace is:
```
System.Net.Tests.AuthenticationTests.NtlmAuthentication_NoRequestHeaders_ReturnsForbiddenStatusCode [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\src\System\Net\Windows\HttpListener.Window
  s.cs(1918,0): at System.Net.HttpListener.DisconnectAsyncResult..ctor(HttpListener httpListener, UInt64 connectionId)
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\src\System\Net\Windows\HttpListener.Window
  s.cs(1645,0): at System.Net.HttpListener.RegisterForDisconnectNotification(UInt64 connectionId, DisconnectAsyncResult
  & disconnectResult)
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\src\System\Net\Windows\HttpListener.Window
  s.cs(1306,0): at System.Net.HttpListener.HandleAuthentication(RequestContextBase memoryBlob, Boolean& stoleBlob)
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\src\System\Net\Windows\HttpListener.Window
  s.cs(655,0): at System.Net.HttpListener.GetContext()
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\tests\AuthenticationTests.cs(310,0): at Sy
  stem.Net.Tests.AuthenticationTests.<AuthenticationFailure>b__30_0()
              at System.Threading.Tasks.Task`1.InnerInvoke()
              at System.Threading.Tasks.Task.<>c.<.cctor>b__276_1(Object obj)
              at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Obj
  ect state)
              at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\tests\AuthenticationTests.cs(317,0): at Sy
  stem.Net.Tests.AuthenticationTests.<AuthenticationFailure>d__30.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
           C:\Users\hugh\Documents\GitHub\corefx\src\System.Net.HttpListener\tests\AuthenticationTests.cs(127,0): at Sy
  stem.Net.Tests.AuthenticationTests.<NtlmAuthentication_NoRequestHeaders_ReturnsForbiddenStatusCode>d__14.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
```

HttpListener... the gift that keeps on giving

**Comment by hughbe** [comment]
This also happens with Negotiate authentication with the same stack trace

**Comment by karelz** [comment]
AFAIK I remember this was by design porting choice as we ran out of time.
@davidsh @CIPop @stephentoub can you please confirm?

If it is correct, we will wait for feedback on the missing functionality, we will address it in future versions. We should mention it in [ApiCompat](https://github.com/dotnet/corefx/wiki/ApiCompat).

**Comment by stephentoub** [comment]
I'm not aware of any reason this should fail on Windows.  On Unix, yes, we punted on NTLM/Negotiate.

**Comment by karelz** [comment]
@davidsh @CIPop @DavidGoll can you please take a look at this bug?

**Comment by CIPop** [comment]
I don't think `NullReferenceException` is something expected. Sounds like an implementation bug.

**Comment by karelz** [comment]
Triage: Let's fix `NullReferenceException` in 2.0.
@CIPop to investigate the failure after (Stephen tried it) - we may punt that part. Please provide recommendation.

**Comment by CIPop** [comment]
@karelz @DavidGoll @stephentoub 

I have a fix for the NRE but I found other related issues:

1. The HttpListener implementation in CoreFX is different from .Net Desktop because of the new IOCP/ThreadPool model. (Not properly initializing the threadpool bound handle is what was causing the NRE.)
2. The `IsWindowsImplementation` test helper (used to selectively run tests) was returning `false` for NetFX. I've fixed that and now I see that some of the tests fail on Desktop too.

My recommendation is to fix the NRE in 2.0 - that's a clear bug (I'll submit a PR but keep the tests disabled.)
I'll need more time to understand and stabilize the tests in CoreFX vs NetFX and figure out if we have more bugs.

**Comment by CIPop** [comment]
One more thing that I'm not sure of: `AuthenticationFailure` seems to have a race condition, making assumptions on which task completes first:

```C#
            // The client task should complete first - the server should send a 401 response.
            Task resultTask = await Task.WhenAny(clientTask, serverTask);
            if (resultTask == serverTask)
            {
                await serverTask;
            }

            Assert.Same(clientTask, resultTask);
```

Since `clientTask` and `serverTask` are not chained via `ContinueWith` I don't see how the assumption can be guaranteed to always be true.

**Comment by karelz** [comment]
Thanks @CIPop for the analysis, let's get the NRE fix in.
If you think you found another clear bug in `AuthenticationFailure` which is likely going to bite us, please submit fix as well. A separate one.

Let's continue deeper investigation of the post-`IsWindowsImplementation` fix in parallel. I expect that will slip into Escrow. If we find more bugs later, we can decide later what to take in 2.0 / 2.0.x servicing.

**Comment by stephentoub** [comment]
> One more thing that I'm not sure of

See my response here: https://github.com/dotnet/corefx/pull/20341/files#r118621111

**Comment by CIPop** [comment]
NRE is fixed by #20374 (2.0 backport in #20375). 

Changing the scope of this bug to take a closer look at the tests.

**Comment by hughbe** [comment]
Is this a test-bug? I thought this tracks the hangs in authentication with netcoreapp?

**Comment by stephentoub** [comment]
> I thought this tracks the hangs in authentication with netcoreapp?

When you say hang, you just mean the Task returned by GetContextAsync not completing because nothing was accepted because the request was rejected, right?

**Comment by hughbe** [comment]
Right. There are two tests disabled with this issue that pass with netfx:

- NtlmAuthentication_Conversation_ReturnsExpectedType2Message
- NtlmAuthentication_InvalidRequestHeaders_ReturnsExpectedStatusCode

This is a bit of a blocker for adding tests for authentication

**Comment by stephentoub** [comment]
> Right

Ok, so what do you mean by "I thought this tracks the hangs in authentication"... what hangs?

**Comment by karelz** [comment]
@CIPop can you please move the remaining work (with description) into separate issue?
Then we can close this one as it has been addressed & ported to rel/2.0.0 by Stephen.

**Comment by CIPop** [comment]
@hughbe, @stephentoub I will create a different bug. To answer your questions:

1. At least on my machine, the tests fail on .Net Framework as well. I presume they were not executed because of the issue with `IsWindowsImplementation`/`IsNotWindowsImplementation` (see #20341 for a potential fix in `GetContextHelper.cs`).

2. After applying my version of test fixes, I've figured out that the actual behavior is not a hang but a timeout (it will take each test the default 100s request timeout to complete eventually).

A potential reason for this behavior difference is the way WinHTTP is handling null/empty credentials when performing Negotiate auth to localhost servers (it uses the current user token). We will continue the investigation for 2.1.0.

**Comment by CIPop** [comment]
Test / WinHTTP behavior tracked for 2.1.0 in #20604.
This issue tracks the NRE fix in 2.1.0 and 2.0.0.
